ArrayOutOfBounds in `SlimeWorldFactory#transform` method

Seems like the array of block ids are out of the exactly value.
```
[13:26:40] [Server thread/WARN]: Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 14 out of bounds for length 14

[13:26:40] [Server thread/WARN]: 	at me.luizotavio.minecraft.codec.SlimeOutputStream.getBlocksId(SlimeOutputStream.java:357)

[13:26:40] [Server thread/WARN]: 	at me.luizotavio.minecraft.codec.SlimeOutputStream.write(SlimeOutputStream.java:189)

[13:26:40] [Server thread/WARN]: 	at me.luizotavio.minecraft.factory.BukkitSlimeWorldFactory.transform(BukkitSlimeWorldFactory.java:115)

[13:26:40] [Server thread/WARN]: 	... 24 more
```

--- a/api/src/main/java/me/luizotavio/minecraft/common/factory/SlimeWorldFactory.java
+++ b/api/src/main/java/me/luizotavio/minecraft/common/factory/SlimeWorldFactory.java
@@ -44,7 +44,6 @@ import java.io.File;
  * @since 12/08/2022
  */
 public interface SlimeWorldFactory {
-
     /**
      * Creates a new slime world based on {@link SlimeLoaderStrategy} from {@link SlimeKorld}}
      * Remember, after creating a new slime world, you must call {@link SlimeWorld#initialize()} to initialize the world.
@@ -54,7 +53,6 @@ public interface SlimeWorldFactory {
      * @return The new slime world.
      * @throws InternalSlimeException If an error occurs.
      */
-    @Nullable
     SlimeWorld createWorld(@NotNull String name, @NotNull WorldVersion worldVersion, SettingsProperty<?> ... properties) throws InternalSlimeException;
 
     /**

--- a/api/src/main/java/me/luizotavio/minecraft/common/settings/SettingsProperty.java
+++ b/api/src/main/java/me/luizotavio/minecraft/common/settings/SettingsProperty.java
@@ -38,9 +38,10 @@ public abstract class SettingsProperty<T> implements Cloneable {
 
     /**
      * Create a string property.
-     * @param name The name of the property.
+     *
+     * @param name         The name of the property.
      * @param defaultValue The default value of the property.
-     * @param value The value of the property.
+     * @param value        The value of the property.
      * @return The property.
      */
     public static SettingsProperty<String> createStringProperty(String name, String defaultValue, @Nullable String value) {
@@ -49,12 +50,13 @@ public abstract class SettingsProperty<T> implements Cloneable {
 
     /**
      * Create an enum property.
-     * @param name The name of the property.
+     *
+     * @param name         The name of the property.
      * @param defaultValue The default value of the property.
-     * @param value The value of the property.
-     * @param clazz The class of the enum.
+     * @param value        The value of the property.
+     * @param clazz        The class of the enum.
+     * @param <T>          The type of the enum.
      * @return The property.
-     * @param <T> The type of the enum.
      */
     public static <T extends Enum<T>> SettingsProperty<T> createEnumProperty(String name, T defaultValue, @Nullable T value, @NotNull Class<T> clazz) {
         return new EnumProperty<>(name, value == null ? defaultValue : value, defaultValue, clazz);
@@ -62,12 +64,13 @@ public abstract class SettingsProperty<T> implements Cloneable {
 
     /**
      * Create an array based property.
-     * @param name The name of the property.
+     *
+     * @param name         The name of the property.
      * @param defaultValue The default value of the property.
-     * @param value The value of the property.
-     * @param clazz The class of the array.
+     * @param value        The value of the property.
+     * @param clazz        The class of the array.
+     * @param <T>          The type of the array.
      * @return The property.
-     * @param <T> The type of the array.
      */
     public static <T> SettingsProperty<T[]> createArrayProperty(String name, T[] defaultValue, @Nullable T[] value, @NotNull Class<T> clazz) {
         return new ArrayProperty<>(name, value == null ? defaultValue : value, defaultValue, clazz);
@@ -75,9 +78,10 @@ public abstract class SettingsProperty<T> implements Cloneable {
 
     /**
      * Create an integer property.
-     * @param name The name of the property.
+     *
+     * @param name         The name of the property.
      * @param defaultValue The default value of the property.
-     * @param value The value of the property.
+     * @param value        The value of the property.
      * @return The property.
      */
     public static SettingsProperty<Integer> createIntegerProperty(String name, int defaultValue, int value) {
@@ -86,9 +90,10 @@ public abstract class SettingsProperty<T> implements Cloneable {
 
     /**
      * Create a boolean property.
-     * @param name The name of the property.
+     *
+     * @param name         The name of the property.
      * @param defaultValue The default value of the property.
-     * @param value The value of the property.
+     * @param value        The value of the property.
      * @return The property.
      */
     public static SettingsProperty<Boolean> createBooleanProperty(String name, boolean defaultValue, boolean value) {
@@ -97,6 +102,7 @@ public abstract class SettingsProperty<T> implements Cloneable {
 
     /**
      * Retrieve the name of the property.
+     *
      * @return The name of the property.
      */
     @NotNull
@@ -104,13 +110,22 @@ public abstract class SettingsProperty<T> implements Cloneable {
 
     /**
      * Retrieve the current value of the property.
+     *
      * @return The current value of the property.
      */
     @NotNull
     public abstract T getValue();
 
     /**
+     * Update the value of the property.
+     *
+     * @param value The new value of the property.
+     */
+    public abstract void setValue(@NotNull T value);
+
+    /**
      * Retrieve the default value of the property.
+     *
      * @return The default value of the property.
      */
     @NotNull
@@ -142,7 +157,7 @@ public abstract class SettingsProperty<T> implements Cloneable {
     public static class BooleanProperty extends SettingsProperty<Boolean> {
 
         private final String name;
-        private final Boolean value;
+        private Boolean value;
         private final Boolean defaultValue;
 
         public BooleanProperty(String name, Boolean value, Boolean defaultValue) {
@@ -162,7 +177,12 @@ public abstract class SettingsProperty<T> implements Cloneable {
         }
 
         @Override
-        public Boolean getDefaultValue() {
+        public void setValue(@NotNull Boolean value) {
+            this.value = value;
+        }
+
+        @Override
+        public @NotNull Boolean getDefaultValue() {
             return defaultValue;
         }
 
@@ -171,7 +191,7 @@ public abstract class SettingsProperty<T> implements Cloneable {
     public static class StringProperty extends SettingsProperty<String> {
 
         private final String name;
-        private final String value;
+        private String value;
         private final String defaultValue;
 
         public StringProperty(String name, String value, String defaultValue) {
@@ -191,7 +211,12 @@ public abstract class SettingsProperty<T> implements Cloneable {
         }
 
         @Override
-        public String getDefaultValue() {
+        public void setValue(@NotNull String value) {
+            this.value = value;
+        }
+
+        @Override
+        public @NotNull String getDefaultValue() {
             return defaultValue;
         }
 
@@ -200,7 +225,7 @@ public abstract class SettingsProperty<T> implements Cloneable {
     public static class IntegerProperty extends SettingsProperty<Integer> {
 
         private final String name;
-        private final Integer value;
+        private Integer value;
         private final Integer defaultValue;
 
         public IntegerProperty(String name, Integer value, Integer defaultValue) {
@@ -220,7 +245,12 @@ public abstract class SettingsProperty<T> implements Cloneable {
         }
 
         @Override
-        public Integer getDefaultValue() {
+        public void setValue(@NotNull Integer value) {
+            this.value = value;
+        }
+
+        @Override
+        public @NotNull Integer getDefaultValue() {
             return defaultValue;
         }
     }
@@ -228,7 +258,7 @@ public abstract class SettingsProperty<T> implements Cloneable {
     public static class EnumProperty<E extends Enum<E>> extends SettingsProperty<E> {
 
         private final String name;
-        private final E value;
+        private E value;
         private final E defaultValue;
         private final Class<E> enumClass;
 
@@ -250,7 +280,12 @@ public abstract class SettingsProperty<T> implements Cloneable {
         }
 
         @Override
-        public E getDefaultValue() {
+        public void setValue(@NotNull E value) {
+            this.value = value;
+        }
+
+        @Override
+        public @NotNull E getDefaultValue() {
             return defaultValue;
         }
 
@@ -262,7 +297,7 @@ public abstract class SettingsProperty<T> implements Cloneable {
     public static class ArrayProperty<T> extends SettingsProperty<T[]> {
 
         private final String name;
-        private final T[] value;
+        private T[] value;
         private final T[] defaultValue;
         private final Class<T> arrayClass;
 
@@ -284,7 +319,12 @@ public abstract class SettingsProperty<T> implements Cloneable {
         }
 
         @Override
-        public T[] getDefaultValue() {
+        public void setValue(T @NotNull [] value) {
+            this.value = value;
+        }
+
+        @Override
+        public T @NotNull [] getDefaultValue() {
             return defaultValue;
         }
 

--- a/api/src/main/java/me/luizotavio/minecraft/common/settings/factory/SettingsPropertyFactory.java
+++ b/api/src/main/java/me/luizotavio/minecraft/common/settings/factory/SettingsPropertyFactory.java
@@ -47,5 +47,10 @@ public class SettingsPropertyFactory {
     public static final SettingsProperty<Difficulty> DIFFICULTY = SettingsProperty.createEnumProperty("difficulty", Difficulty.NORMAL, Difficulty.NORMAL, Difficulty.class);
     public static final SettingsProperty<GameMode> GAMEMODE = SettingsProperty.createEnumProperty("gamemode", GameMode.SURVIVAL, GameMode.SURVIVAL, GameMode.class);
 
-    public static final SettingsProperty<Integer[]> SPAWN_LOCATION = SettingsProperty.createArrayProperty("spawnLocation", new Integer[3], new Integer[3], Integer.class);
+    public static final SettingsProperty<Integer[]> SPAWN_LOCATION = SettingsProperty.createArrayProperty(
+        "spawnLocation",
+        new Integer[] { 0, 50, 0 },
+        new Integer[] { 0, 50, 0 },
+        Integer.class
+    );
 }

--- a/api/src/main/java/me/luizotavio/minecraft/common/settings/factory/SettingsPropertyFactory.java
+++ b/api/src/main/java/me/luizotavio/minecraft/common/settings/factory/SettingsPropertyFactory.java
@@ -37,12 +37,12 @@ import org.bukkit.GameMode;
  */
 public class SettingsPropertyFactory {
 
-    public static final SettingsProperty<Boolean> HAS_ENTITIES = SettingsProperty.createBooleanProperty("hasEntities", false, true);
-    public static final SettingsProperty<Boolean> HAS_EXTRA_DATA = SettingsProperty.createBooleanProperty("hasExtraData", false, true);
-    public static final SettingsProperty<Boolean> HAS_PVP = SettingsProperty.createBooleanProperty("hasPVP", false, true);
+    public static final SettingsProperty<Boolean> HAS_ENTITIES = SettingsProperty.createBooleanProperty("hasEntities", false, false);
+    public static final SettingsProperty<Boolean> HAS_EXTRA_DATA = SettingsProperty.createBooleanProperty("hasExtraData", false, false);
+    public static final SettingsProperty<Boolean> HAS_PVP = SettingsProperty.createBooleanProperty("hasPVP", false, false);
     public static final SettingsProperty<Boolean> HAS_MONSTERS = SettingsProperty.createBooleanProperty("hasMonsters", false, true);
     public static final SettingsProperty<Boolean> HAS_ANIMALS = SettingsProperty.createBooleanProperty("hasAnimals", false, true);
-    public static final SettingsProperty<Boolean> SHOULD_SAVE = SettingsProperty.createBooleanProperty("shouldSave", false, true);
+    public static final SettingsProperty<Boolean> SHOULD_SAVE = SettingsProperty.createBooleanProperty("shouldSave", false, false);
 
     public static final SettingsProperty<Difficulty> DIFFICULTY = SettingsProperty.createEnumProperty("difficulty", Difficulty.NORMAL, Difficulty.NORMAL, Difficulty.class);
     public static final SettingsProperty<GameMode> GAMEMODE = SettingsProperty.createEnumProperty("gamemode", GameMode.SURVIVAL, GameMode.SURVIVAL, GameMode.class);

--- a/bukkit/src/main/java/me/luizotavio/minecraft/BukkitSlimeWorld.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/BukkitSlimeWorld.java
@@ -42,9 +42,7 @@ import me.luizotavio.minecraft.data.container.BukkitSlimePersistentContainer;
 import me.luizotavio.minecraft.prototype.ProtoSlimeFile;
 import me.luizotavio.minecraft.world.CraftSlimeWorld;
 import me.luizotavio.minecraft.world.data.SlimeDataManager;
-import net.minecraft.server.v1_8_R3.MinecraftServer;
-import net.minecraft.server.v1_8_R3.NBTTagCompound;
-import net.minecraft.server.v1_8_R3.WorldServer;
+import net.minecraft.server.v1_8_R3.*;
 import org.bukkit.Bukkit;
 import org.bukkit.World;
 import org.bukkit.craftbukkit.v1_8_R3.CraftWorld;
@@ -233,6 +231,19 @@ public class BukkitSlimeWorld implements SlimeWorld {
             for (AbstractSlimeData slimeData : registry.getRegistered()) {
                 slimeData.deserialize(this, persistentContainer);
             }
+        }
+
+        // Fix world maps
+        NBTTagCompound worldMaps = protoSlimeFile.getWorldMaps();
+        NBTTagList maps = worldMaps.getList("maps", 10);
+
+        for (int i = 0; i < maps.size(); i++) {
+            NBTTagCompound map = maps.get(i);
+
+            WorldMap worldMap = new WorldMap("map_" + map.getInt("id"));
+            worldMap.a(map);
+
+            craftWorld.worldMaps.a(worldMap.id, worldMap);
         }
 
         return craftWorld.getWorld();

--- a/bukkit/src/main/java/me/luizotavio/minecraft/BukkitSlimeWorld.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/BukkitSlimeWorld.java
@@ -27,6 +27,7 @@ package me.luizotavio.minecraft;
 import com.google.common.collect.ImmutableSet;
 import de.tr7zw.nbtapi.NBTContainer;
 import me.luizotavio.minecraft.codec.SlimeInputStream;
+import me.luizotavio.minecraft.codec.SlimeOutputStream;
 import me.luizotavio.minecraft.common.SlimeWorld;
 import me.luizotavio.minecraft.common.data.AbstractSlimeData;
 import me.luizotavio.minecraft.common.data.registry.SlimeDataRegistry;
@@ -36,6 +37,7 @@ import me.luizotavio.minecraft.common.event.impl.SlimeWorldUnloadEvent;
 import me.luizotavio.minecraft.common.exception.InternalSlimeException;
 import me.luizotavio.minecraft.common.service.SlimeKorld;
 import me.luizotavio.minecraft.common.settings.SettingsProperty;
+import me.luizotavio.minecraft.common.settings.factory.SettingsPropertyFactory;
 import me.luizotavio.minecraft.common.strategy.SlimeLoaderStrategy;
 import me.luizotavio.minecraft.common.version.WorldVersion;
 import me.luizotavio.minecraft.data.container.BukkitSlimePersistentContainer;
@@ -50,6 +52,7 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
 import java.util.Hashtable;
 import java.util.Map;
 import java.util.Set;
@@ -161,6 +164,21 @@ public class BukkitSlimeWorld implements SlimeWorld {
             throw new InternalSlimeException("World unload cancelled");
         }
 
+        if (getProperty(SettingsPropertyFactory.SHOULD_SAVE)) {
+            SlimeKorld slimeKorld = getKorld();
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+
+            try (SlimeOutputStream slimeOutputStream = new SlimeOutputStream(outputStream, this, slimeKorld.getDataRegistry())) {
+                slimeOutputStream.write();
+            } catch (Exception e) {
+                throw new InternalSlimeException("Failed to save world data", e);
+            }
+
+            byte[] data = outputStream.toByteArray();
+
+            slimeKorld.getLoaderStrategy().save(this, data);
+        }
+
         Bukkit.unloadWorld(world, true);
     }
 
@@ -174,6 +192,10 @@ public class BukkitSlimeWorld implements SlimeWorld {
             .getLoaderStrategy();
 
         byte[] data = loader.load(name, false);
+
+        if (data == null) {
+            throw new InternalSlimeException("World data not found");
+        }
 
         ProtoSlimeFile protoSlimeFile;
 

--- a/bukkit/src/main/java/me/luizotavio/minecraft/BukkitSlimeWorld.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/BukkitSlimeWorld.java
@@ -115,7 +115,18 @@ public class BukkitSlimeWorld implements SlimeWorld {
 
     @Override
     public <T> void setProperty(@NotNull SettingsProperty<T> property, @NotNull T value) {
-        properties.put(property.getName(), property.clone());
+        SettingsProperty current = properties.get(property.getName());
+
+        if (current != null) {
+            current.setValue(value);
+        } else {
+            SettingsProperty clone = property.clone();
+
+            // Fixed the bug of the property not being cloned and updated with the new value
+            clone.setValue(value);
+
+            properties.put(property.getName(), clone);
+        }
     }
 
     @Override
@@ -177,6 +188,7 @@ public class BukkitSlimeWorld implements SlimeWorld {
             byte[] data = outputStream.toByteArray();
 
             slimeKorld.getLoaderStrategy().save(this, data);
+            System.out.println("saving world data");
         }
 
         Bukkit.unloadWorld(world, true);

--- a/bukkit/src/main/java/me/luizotavio/minecraft/codec/SlimeInputStream.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/codec/SlimeInputStream.java
@@ -39,6 +39,7 @@ import java.util.Arrays;
 import java.util.BitSet;
 import java.util.Set;
 
+import static me.luizotavio.minecraft.common.version.SlimeVersion.CURRENT_SLIME_VERSION;
 import static me.luizotavio.minecraft.common.version.SlimeVersion.SLIME_MAGIC_HEADER;
 import static org.apache.commons.lang.ArrayUtils.EMPTY_BYTE_ARRAY;
 
@@ -73,6 +74,13 @@ public class SlimeInputStream extends DataInputStream {
 
         if (!Arrays.equals(SLIME_MAGIC_HEADER, magic)) {
             throw new InternalSlimeException("Invalid magic header");
+        }
+
+        // Fix slime version
+        byte slimeVersion = readByte();
+
+        if (slimeVersion != CURRENT_SLIME_VERSION) {
+            throw new InternalSlimeException("That slime version isn't supported.");
         }
 
         byte version = readByte();

--- a/bukkit/src/main/java/me/luizotavio/minecraft/codec/SlimeInputStream.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/codec/SlimeInputStream.java
@@ -25,11 +25,15 @@
 package me.luizotavio.minecraft.codec;
 
 import com.github.luben.zstd.Zstd;
+import com.google.common.io.ByteStreams;
 import me.luizotavio.minecraft.common.exception.InternalSlimeException;
 import me.luizotavio.minecraft.common.settings.SettingsProperty;
-import me.luizotavio.minecraft.common.settings.factory.SettingsPropertyFactory;
 import me.luizotavio.minecraft.common.version.WorldVersion;
 import me.luizotavio.minecraft.prototype.ProtoSlimeFile;
+import net.minecraft.server.v1_8_R3.NBTCompressedStreamTools;
+import net.minecraft.server.v1_8_R3.NBTReadLimiter;
+import net.minecraft.server.v1_8_R3.NBTTagCompound;
+import net.minecraft.server.v1_8_R3.NBTTagList;
 import org.jetbrains.annotations.NotNull;
 
 import java.io.DataInputStream;
@@ -41,7 +45,6 @@ import java.util.Set;
 
 import static me.luizotavio.minecraft.common.version.SlimeVersion.CURRENT_SLIME_VERSION;
 import static me.luizotavio.minecraft.common.version.SlimeVersion.SLIME_MAGIC_HEADER;
-import static org.apache.commons.lang.ArrayUtils.EMPTY_BYTE_ARRAY;
 
 /**
  * @author Luiz Otávio de Farias Corrêa
@@ -105,7 +108,7 @@ public class SlimeInputStream extends DataInputStream {
             throw new InternalSlimeException("Invalid depth or width");
         }
 
-        byte[] populatedChunks = new byte[(width * depth) >> 3];
+        byte[] populatedChunks = new byte[(int) Math.ceil((width * depth) / 8.0D)];
 
         success = read(populatedChunks);
 
@@ -120,12 +123,12 @@ public class SlimeInputStream extends DataInputStream {
 
         boolean hasEntities = readBoolean();
 
-        byte[] uncompressedEntities;
+        NBTTagList uncompressedEntities;
 
-        if (hasEntities && properties.contains(SettingsPropertyFactory.HAS_ENTITIES)) {
-            uncompressedEntities = readCompressed();
+        if (hasEntities) {
+            uncompressedEntities = readList(readCompressed(), "entities");
         } else {
-            uncompressedEntities = EMPTY_BYTE_ARRAY;
+            uncompressedEntities = new NBTTagList();
         }
 
         byte[] extraData = readCompressed(),
@@ -145,10 +148,10 @@ public class SlimeInputStream extends DataInputStream {
             minZ,
             bitSet,
             uncompressedChunks,
-            uncompressedTiles,
-            isExtra ? mapData : extraData,
+            readList(uncompressedTiles, "tiles"),
             uncompressedEntities,
-            isExtra ? extraData : null
+            isExtra ? readCompound(extraData) : null,
+            !isExtra ? readCompound(extraData) : null
         );
     }
 
@@ -165,5 +168,14 @@ public class SlimeInputStream extends DataInputStream {
         }
 
         return Zstd.decompress(compressed, uncompressedSize);
+    }
+
+    private NBTTagCompound readCompound(byte[] src) throws IOException {
+        return NBTCompressedStreamTools.a(ByteStreams.newDataInput(src), NBTReadLimiter.a);
+    }
+
+    private NBTTagList readList(byte[] src, String key) throws IOException {
+        NBTTagCompound compound = readCompound(src);
+        return compound.getList(key, 10);
     }
 }

--- a/bukkit/src/main/java/me/luizotavio/minecraft/codec/SlimeOutputStream.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/codec/SlimeOutputStream.java
@@ -59,12 +59,12 @@ public class SlimeOutputStream extends DataOutputStream {
 
     public static final WorldVersion SUPPORTED_VERSION = WorldVersion.V1_8_R3;
 
-    public static final int CHUNK_SIZE = 2 ^ 14;
-    public static final int NIBBLE_ARRAY_SIZE = 2 ^ 11;
-    public static final int BLOCK_DATA_SIZE = 2 ^ 12;
-    public static final int SECTION_PER_CHUNK = 4 ^ 2;
-    public static final int HEIGHTMAP_SIZE = 2 ^ 8;
-    public static final int BIOME_SIZE = 2 ^ 8;
+    public static final int CHUNK_SIZE = 16384;
+    public static final int NIBBLE_ARRAY_SIZE = 2048;
+    public static final int BLOCK_DATA_SIZE = 4096;
+    public static final int SECTION_PER_CHUNK = 16;
+    public static final int HEIGHTMAP_SIZE = 256;
+    public static final int BIOME_SIZE = 256;
 
     private final SlimeWorld slimeWorld;
     private final SlimeDataRegistry dataRegistry;

--- a/bukkit/src/main/java/me/luizotavio/minecraft/codec/SlimeOutputStream.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/codec/SlimeOutputStream.java
@@ -142,10 +142,10 @@ public class SlimeOutputStream extends DataOutputStream {
         BitSet bitSet = new BitSet(width * depth);
 
         for (Chunk chunk : chunks) {
-            bitSet.set(chunk.locX - minX + (chunk.locZ - minZ) * width, true);
+            bitSet.set((chunk.locZ - minZ) * width + (chunk.locX - minX) , true);
         }
 
-        int chunkSize = (int) Math.ceil((width * depth) >> 3);
+        int chunkSize = (int) Math.ceil((width * depth) / 8.0D);
 
         fromBitSet(this, bitSet, chunkSize);
 
@@ -188,13 +188,17 @@ public class SlimeOutputStream extends DataOutputStream {
 
                     Pair<byte[], NibbleArray> blockData = getBlocksId(chunkSection);
 
-                    write(blockData.getKey());
-                    write(blockData.getValue().a());
+                    dataOutputStream.write(blockData.getKey());
+                    dataOutputStream.write(blockData.getValue().a());
 
                     byte[] skyLight = chunkSection.getSkyLightArray()
                         .a();
 
                     dataOutputStream.writeBoolean(skyLight != null && skyLight.length == NIBBLE_ARRAY_SIZE);
+
+                    if (skyLight != null) {
+                        dataOutputStream.write(skyLight);
+                    }
                 }
             }
         }
@@ -238,7 +242,9 @@ public class SlimeOutputStream extends DataOutputStream {
 
         write(tileCompressed);
 
+        // Fix the boolean of has entities
         if (slimeWorld.hasProperty(SettingsPropertyFactory.HAS_ENTITIES)) {
+            writeBoolean(true);
             NBTTagList entityTagList = new NBTTagList();
 
             for (Chunk chunk : chunks) {
@@ -267,6 +273,8 @@ public class SlimeOutputStream extends DataOutputStream {
             writeInt(entityBytes.length); // Not compressed size
 
             write(entityCompressed);
+        } else {
+            writeBoolean(false);
         }
 
         if (slimeWorld.hasProperty(SettingsPropertyFactory.HAS_EXTRA_DATA)) {
@@ -338,7 +346,7 @@ public class SlimeOutputStream extends DataOutputStream {
 
         stream.write(bytes);
 
-        int padding = chunkSize - bitSet.size();
+        int padding = chunkSize - bytes.length;
 
         for (int i = 0; i < padding; i++) {
             stream.write(0);

--- a/bukkit/src/main/java/me/luizotavio/minecraft/factory/BukkitSlimeWorldFactory.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/factory/BukkitSlimeWorldFactory.java
@@ -56,7 +56,7 @@ public class BukkitSlimeWorldFactory implements SlimeWorldFactory {
     }
 
     @Override
-    public @Nullable SlimeWorld createWorld(@NotNull String name, @NotNull WorldVersion worldVersion, SettingsProperty<?>... properties) throws InternalSlimeException {
+    public SlimeWorld createWorld(@NotNull String name, @NotNull WorldVersion worldVersion, SettingsProperty<?>... properties) throws InternalSlimeException {
         World world = Bukkit.getWorld(name);
 
         if (world != null) {

--- a/bukkit/src/main/java/me/luizotavio/minecraft/prototype/ProtoSlimeFile.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/prototype/ProtoSlimeFile.java
@@ -49,11 +49,12 @@ public class ProtoSlimeFile {
     private final WorldVersion worldVersion;
 
     private final int width, depth, minX, minZ;
-
     private final BitSet chunks;
 
-    private final byte[] chunkData, tileData, mapData;
-    private final byte[] entityData, extraData;
+    private final byte[] chunkData;
+
+    private final NBTTagList tileEntities, entities;
+    private final NBTTagCompound extraData, mapData;
 
     public ProtoSlimeFile(
         @NotNull WorldVersion worldVersion,
@@ -63,10 +64,10 @@ public class ProtoSlimeFile {
         int minZ,
         @NotNull BitSet chunks,
         byte[] chunkData,
-        byte[] tileData,
-        byte[] mapData,
-        byte[] entityData,
-        byte[] extraData
+        NBTTagList tileData,
+        NBTTagList entityData,
+        NBTTagCompound extraData,
+        NBTTagCompound mapData
     ) {
         this.worldVersion = worldVersion;
         this.width = width;
@@ -75,10 +76,10 @@ public class ProtoSlimeFile {
         this.minZ = minZ;
         this.chunks = chunks;
         this.chunkData = chunkData;
-        this.tileData = tileData;
         this.mapData = mapData;
-        this.entityData = entityData;
         this.extraData = extraData;
+        this.tileEntities = tileData;
+        this.entities = entityData;
     }
 
     /**
@@ -88,9 +89,6 @@ public class ProtoSlimeFile {
      * @throws IOException If an error occurs while reading the file.
      */
     public TLongObjectHashMap<ProtoSlimeChunk> getProtoChunks() throws IOException {
-        NBTTagCompound entities = readCompound(entityData),
-            tiles = readCompound(tileData);
-
         TLongObjectHashMap<ProtoSlimeChunk> protoChunks = new TLongObjectHashMap<>();
 
         ByteArrayDataInput input = ByteStreams.newDataInput(chunkData);
@@ -113,14 +111,14 @@ public class ProtoSlimeFile {
 
             input.readFully(biomeData);
 
-            byte[] sections = new byte[SECTION_PER_CHUNK >> 3];
+            byte[] sections = new byte[SECTION_PER_CHUNK / 8];
 
             input.readFully(sections);
 
             BitSet populatedSections = BitSet.valueOf(sections);
             ChunkSection[] chunkSections = new ChunkSection[SECTION_PER_CHUNK];
 
-            for (int sectionIndex = 0; sectionIndex != SECTION_PER_CHUNK; sectionIndex++) {
+            for (int sectionIndex = 0; sectionIndex <= SECTION_PER_CHUNK; sectionIndex++) {
                 if (!populatedSections.get(sectionIndex)) {
                     continue;
                 }
@@ -147,7 +145,7 @@ public class ProtoSlimeFile {
                     input.readFully(skyLight);
                 }
 
-                ChunkSection chunkSection = new ChunkSection(sectionIndex, true);
+                ChunkSection chunkSection = new ChunkSection(sectionIndex << 4, true);
 
                 chunkSection.a(
                     new NibbleArray(blockLight)
@@ -174,12 +172,9 @@ public class ProtoSlimeFile {
             protoChunks.put(LongHash.toLong(x, z), protoChunk);
         }
 
-        NBTTagList entitiesList = entities.getList("entities", 10),
-            tilesList = tiles.getList("tiles", 10);
-
-        if (!entitiesList.isEmpty()) {
-            for (int index = 0; index < entitiesList.size(); index++) {
-                NBTTagCompound entity = entitiesList.get(index);
+        if (!entities.isEmpty()) {
+            for (int index = 0; index < entities.size(); index++) {
+                NBTTagCompound entity = entities.get(index);
 
                 NBTTagList position = entity.getList("Pos", 6);
 
@@ -194,9 +189,9 @@ public class ProtoSlimeFile {
             }
         }
 
-        if(!tilesList.isEmpty()) {
-            for (int index = 0; index < tilesList.size(); index++) {
-                NBTTagCompound tile = tilesList.get(index);
+        if(!tileEntities.isEmpty()) {
+            for (int index = 0; index < tileEntities.size(); index++) {
+                NBTTagCompound tile = tileEntities.get(index);
 
                 int x = tile.getInt("x"),
                     z = tile.getInt("z");
@@ -212,13 +207,6 @@ public class ProtoSlimeFile {
         return protoChunks;
     }
 
-    public byte[] getTileData() {
-        return tileData;
-    }
-
-    public byte[] getEntityData() {
-        return entityData;
-    }
 
     public byte[] getChunkData() {
         return chunkData;
@@ -248,30 +236,12 @@ public class ProtoSlimeFile {
         return worldVersion;
     }
 
-    @NotNull
     public NBTTagCompound getWorldMaps() {
-        try {
-            return readCompound(mapData);
-        } catch (IOException e) {
-            throw new RuntimeException(e);
-        }
+        return mapData;
     }
 
     @Nullable
     public NBTTagCompound getExtraData() {
-        if (mapData == null) {
-            return null;
-        }
-
-        try {
-            return readCompound(extraData);
-        } catch (IOException e) {
-            throw new RuntimeException(e);
-        }
+        return extraData;
     }
-
-    private NBTTagCompound readCompound(byte[] src) throws IOException {
-        return NBTCompressedStreamTools.a(ByteStreams.newDataInput(src), NBTReadLimiter.a);
-    }
-
 }

--- a/bukkit/src/main/java/me/luizotavio/minecraft/prototype/ProtoSlimeFile.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/prototype/ProtoSlimeFile.java
@@ -212,10 +212,6 @@ public class ProtoSlimeFile {
         return protoChunks;
     }
 
-    public byte[] getMapData() {
-        return mapData;
-    }
-
     public byte[] getTileData() {
         return tileData;
     }
@@ -250,6 +246,15 @@ public class ProtoSlimeFile {
 
     public WorldVersion getWorldVersion() {
         return worldVersion;
+    }
+
+    @NotNull
+    public NBTTagCompound getWorldMaps() {
+        try {
+            return readCompound(mapData);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
     }
 
     @Nullable

--- a/bukkit/src/main/java/me/luizotavio/minecraft/prototype/chunk/ProtoSlimeChunk.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/prototype/chunk/ProtoSlimeChunk.java
@@ -119,9 +119,4 @@ public class ProtoSlimeChunk {
 
         return chunk;
     }
-
-
-
-
-
 }

--- a/bukkit/src/main/java/me/luizotavio/minecraft/world/CraftSlimeWorld.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/world/CraftSlimeWorld.java
@@ -90,9 +90,6 @@ public class CraftSlimeWorld extends WorldServer {
         );
 
         this.pvpMode = slimeWorld.getProperty(SettingsPropertyFactory.HAS_PVP);
-
-        SlimeChunkLoader slimeChunkLoader = (SlimeChunkLoader) idatamanager.createChunkLoader(worldProvider);
-        slimeChunkLoader.loadAll(this);
     }
 
     @Override

--- a/bukkit/src/main/java/me/luizotavio/minecraft/world/chunkloader/SlimeChunkLoader.java
+++ b/bukkit/src/main/java/me/luizotavio/minecraft/world/chunkloader/SlimeChunkLoader.java
@@ -40,14 +40,30 @@ public class SlimeChunkLoader implements IChunkLoader {
 
     private final TLongObjectHashMap<ProtoSlimeChunk> chunks;
 
+    private final TLongObjectHashMap<Chunk> loadedChunks = new TLongObjectHashMap<>();
+
     public SlimeChunkLoader(TLongObjectHashMap<ProtoSlimeChunk> chunks) {
         this.chunks = chunks;
     }
 
     @Override
     public Chunk a(World world, int i, int i1) throws IOException {
-        return ((WorldServer) world).chunkProviderServer
-            .emptyChunk;
+        long hash = LongHash.toLong(i, i1);
+
+        if (loadedChunks.containsKey(hash)) {
+            return loadedChunks.get(hash);
+        } else {
+            ProtoSlimeChunk chunk = chunks.get(hash);
+
+            if (chunk == null) {
+                return null;
+            }
+
+            Chunk target = chunk.toChunk(world);
+            loadedChunks.put(hash, target);
+
+            return target;
+        }
     }
 
     /**
@@ -70,25 +86,5 @@ public class SlimeChunkLoader implements IChunkLoader {
 
     @Override
     public void b() {
-    }
-
-    /**
-     * Loads all chunks from the file system.
-     *
-     * @param craftSlimeWorld The world to load the chunk into.
-     */
-    public void loadAll(CraftSlimeWorld craftSlimeWorld) {
-        ChunkProviderServer chunkProviderServer = craftSlimeWorld.chunkProviderServer;
-
-        for (ProtoSlimeChunk chunk : chunks.valueCollection()) {
-            Chunk target = chunk.toChunk(craftSlimeWorld);
-
-            chunkProviderServer.chunks.put(
-                LongHash.toLong(target.locX, target.locZ),
-                target
-            );
-
-            target.addEntities();
-        }
     }
 }

--- a/settings.gradle.kts
+++ b/settings.gradle.kts
@@ -2,6 +2,6 @@ rootProject.name = "rewrite-slime-korld"
 
 include(
     "api",
-    "bukkit"
+    "bukkit",
+    "test-plugin"
 )
-

--- a/test-plugin/build.gradle.kts
+++ b/test-plugin/build.gradle.kts
@@ -1,0 +1,41 @@
+import org.gradle.internal.impldep.org.apache.commons.lang.StringUtils
+
+plugins {
+    id("com.github.johnrengelman.shadow") version "7.1.2"
+}
+
+repositories {
+    maven {
+        name = "elmakers"
+        url = uri("https://maven.elmakers.com/repository/")
+    }
+
+    maven {
+        name = "codemc"
+        url = uri("https://repo.codemc.org/repository/maven-public/")
+    }
+
+    maven {
+        name = "jitpack"
+        url = uri("https://jitpack.io")
+    }
+}
+
+val minecraftVersion = "1.8.8-R0.1-SNAPSHOT"
+
+dependencies {
+    compileOnly("org.spigotmc:spigot:$minecraftVersion")
+
+    implementation(project(":api"))
+    implementation(project(":bukkit"))
+
+    implementation("com.github.SaiintBrisson.command-framework:bukkit:1.3.0")
+}
+
+tasks.getByName("jar") {
+    enabled = false
+}
+
+tasks.shadowJar {
+    archiveClassifier.set(StringUtils.EMPTY)
+}

--- a/test-plugin/src/main/java/me/luizotavio/minecraft/plugin/SlimePlugin.java
+++ b/test-plugin/src/main/java/me/luizotavio/minecraft/plugin/SlimePlugin.java
@@ -1,0 +1,134 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) [2022] [LUIZ O. F. CORRÊA]
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package me.luizotavio.minecraft.plugin;
+
+import me.luizotavio.minecraft.BukkitSlimeKorld;
+import me.luizotavio.minecraft.common.SlimeWorld;
+import me.luizotavio.minecraft.common.service.SlimeKorld;
+import me.luizotavio.minecraft.common.version.WorldVersion;
+import me.luizotavio.minecraft.plugin.command.SlimeCommand;
+import me.luizotavio.minecraft.plugin.listener.SlimeHandler;
+import me.saiintbrisson.bukkit.command.BukkitFrame;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
+import org.bukkit.WorldCreator;
+import org.bukkit.plugin.java.JavaPlugin;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+/**
+ * @author Luiz Otávio de Farias Corrêa
+ * @since 13/08/2022
+ */
+public class SlimePlugin extends JavaPlugin {
+
+    private File worldFolder;
+
+    @Override
+    public void onLoad() {
+        worldFolder = new File(Bukkit.getWorldContainer(), "world-slime-test");
+
+        // Copy the folder to the world folder
+        if (!worldFolder.exists()) {
+            InputStream inputStream = getResource("world-slime-test.zip");
+
+            if (inputStream == null) {
+                throw new RuntimeException("Could not find world-slime-test folder");
+            }
+
+            try {
+                // Unzip the folder
+                try (ZipInputStream stream = new ZipInputStream(inputStream)) {
+                    ZipEntry zipEntry;
+
+                    byte[] bytes = new byte[1024];
+
+                    while ((zipEntry = stream.getNextEntry()) != null) {
+                        File file = new File(Bukkit.getWorldContainer(), zipEntry.getName());
+
+                        if (zipEntry.isDirectory()) {
+                            file.mkdirs();
+                        } else {
+                            try (FileOutputStream outputStream = new FileOutputStream(file)) {
+                                int length;
+
+                                while ((length = stream.read(bytes)) > 0) {
+                                    outputStream.write(bytes, 0, length);
+                                }
+                            }
+                        }
+                    }
+                }
+
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+
+            Bukkit.getLogger().info("World folder created");
+        }
+    }
+
+    @Override
+    public void onEnable() {
+        SlimeKorld slimeKorld = BukkitSlimeKorld.createInstance(this);
+
+        World world = null;
+
+        boolean exists;
+        try {
+            SlimeWorld slimeData = slimeKorld.getFactory()
+                .createWorld("world-slime-test", WorldVersion.V1_8_R3);
+
+            world = slimeData.initialize();
+
+            exists = true;
+
+            Bukkit.getLogger().info("World found");
+        } catch (Exception exception) {
+            exists = false;
+        }
+
+        if (!exists) {
+            world = new WorldCreator("world-slime-test")
+                .createWorld();
+        }
+
+        BukkitFrame bukkitFrame = new BukkitFrame(this);
+        bukkitFrame.registerCommands(new SlimeCommand());
+
+        getServer().getPluginManager().registerEvents(new SlimeHandler(world), this);
+
+        Bukkit.getLogger().info("Plugin enabled");
+    }
+
+    @Override
+    public void onDisable() {
+    }
+}

--- a/test-plugin/src/main/java/me/luizotavio/minecraft/plugin/command/SlimeCommand.java
+++ b/test-plugin/src/main/java/me/luizotavio/minecraft/plugin/command/SlimeCommand.java
@@ -1,0 +1,106 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) [2022] [LUIZ O. F. CORRÊA]
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package me.luizotavio.minecraft.plugin.command;
+
+import me.luizotavio.minecraft.common.SlimeWorld;
+import me.luizotavio.minecraft.common.exception.InternalSlimeException;
+import me.luizotavio.minecraft.common.service.SlimeKorld;
+import me.luizotavio.minecraft.common.settings.factory.SettingsPropertyFactory;
+import me.luizotavio.minecraft.common.util.Pair;
+import me.luizotavio.minecraft.common.version.WorldVersion;
+import me.saiintbrisson.minecraft.command.annotation.Command;
+import me.saiintbrisson.minecraft.command.command.Context;
+import me.saiintbrisson.minecraft.command.target.CommandTarget;
+import org.bukkit.Bukkit;
+import org.bukkit.entity.Player;
+
+import java.io.File;
+
+/**
+ * @author Luiz Otávio de Farias Corrêa
+ * @since 13/08/2022
+ */
+public class SlimeCommand {
+
+    @Command(
+        name = "slime",
+        target = CommandTarget.PLAYER
+    )
+    public void handleSlimeCommand(Context<Player> context) {
+        Player player = context.getSender();
+
+        player.sendMessage(
+            new String[]{
+                "§aSlime world testing version: §e" + Bukkit.getVersion(),
+                "§aType /slime convert - to convert your world to the new format",
+                "§aType /slime save - to save your world"
+            }
+        );
+    }
+
+    @Command(
+        name = "slime.convert"
+    )
+    public void handleSlimeConvertCommand(Context<Player> context) throws InternalSlimeException {
+        Player player = context.getSender();
+
+        player.sendMessage("§aConverting world to new format...");
+
+        SlimeKorld slimeKorld = Bukkit.getServicesManager()
+            .load(SlimeKorld.class);
+
+        Pair<SlimeWorld, byte[]> pair = slimeKorld.getFactory()
+            .transform(player.getWorld());
+
+        // Save it to the new format
+        slimeKorld.getLoaderStrategy().save(pair.getKey(), pair.getValue());
+
+        player.sendMessage("§aWorld converted to new format");
+    }
+
+    @Command(
+        name = "slime.save"
+    )
+    public void handleSlimeSaveCommand(Context<Player> context) throws InternalSlimeException {
+        Player player = context.getSender();
+
+        player.sendMessage("§aSaving world...");
+
+        SlimeKorld slimeKorld = Bukkit.getServicesManager()
+            .load(SlimeKorld.class);
+
+        SlimeWorld slimeWorld  = slimeKorld.getFactory()
+                .createWorld(
+                    player.getWorld(),
+                    WorldVersion.V1_8_R3
+                );
+
+        slimeWorld.setProperty(SettingsPropertyFactory.SHOULD_SAVE, true);
+        slimeWorld.unload();
+
+        player.sendMessage("§aWorld saved");
+    }
+
+}

--- a/test-plugin/src/main/java/me/luizotavio/minecraft/plugin/listener/SlimeHandler.java
+++ b/test-plugin/src/main/java/me/luizotavio/minecraft/plugin/listener/SlimeHandler.java
@@ -1,0 +1,59 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) [2022] [LUIZ O. F. CORRÊA]
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package me.luizotavio.minecraft.plugin.listener;
+
+import org.bukkit.World;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.player.PlayerJoinEvent;
+
+/**
+ * @author Luiz Otávio de Farias Corrêa
+ * @since 13/08/2022
+ */
+public class SlimeHandler implements Listener {
+
+    private final World world;
+
+    public SlimeHandler(World world) {
+        this.world = world;
+    }
+
+    @EventHandler(priority = EventPriority.LOWEST)
+    public void onPlayerJoin(PlayerJoinEvent event) {
+        Player player = event.getPlayer();
+
+        player.setAllowFlight(true);
+        player.setFlying(true);
+
+        if (player.getWorld() == world) {
+            return;
+        }
+
+        player.teleport(world.getSpawnLocation());
+    }
+}

--- a/test-plugin/src/main/resources/plugin.yml
+++ b/test-plugin/src/main/resources/plugin.yml
@@ -1,0 +1,6 @@
+name: SlimeKorldPlugin
+version: 1.8.8-R0.1-SNAPSHOT
+main: me.luizotavio.minecraft.plugin.SlimePlugin
+load: POSTWORLD
+authors:
+    - luiz-otavio

